Error using ssl endpoint

When using the default aws s3 client created by `S3MockRule.createS3Client()`, I'm getting an exception when trying to use the client. It doesn't matter which operation I try, the root exception is the same. Stacktrace:

```
com.amazonaws.SdkClientException: Unable to execute HTTP request: Unrecognized SSL message, plaintext connection?

	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleRetryableException(AmazonHttpClient.java:1116)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1066)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4368)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4315)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1758)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1613)
	at com.example.messaging.RouteTest.testTransform(RouteTest.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:73)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:83)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at com.adobe.testing.s3mock.junit4.S3MockRule$1.evaluate(S3MockRule.java:66)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: javax.net.ssl.SSLException: Unrecognized SSL message, plaintext connection?
	at sun.security.ssl.InputRecord.handleUnknownRecord(InputRecord.java:710)
	at sun.security.ssl.InputRecord.read(InputRecord.java:527)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:983)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1385)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1413)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1397)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:396)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:355)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:373)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.amazonaws.http.conn.ClientConnectionManagerFactory$Handler.invoke(ClientConnectionManagerFactory.java:76)
	at com.amazonaws.http.conn.$Proxy119.connect(Unknown Source)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:381)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:237)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at com.amazonaws.http.apache.client.impl.SdkHttpClient.execute(SdkHttpClient.java:72)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1238)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
	... 43 more
```

Switching to creating an s3 client like so:
```
final BasicAWSCredentials credentials = new BasicAWSCredentials("foo", "bar");

return AmazonS3ClientBuilder.standard()
        .withCredentials(new AWSStaticCredentialsProvider(credentials))
        .withClientConfiguration(
                S3_MOCK_RULE.configureClientToIgnoreInvalidSslCertificates(new ClientConfiguration()))
        .withEndpointConfiguration(
                new AwsClientBuilder.EndpointConfiguration("http://localhost:" + S3_MOCK_RULE.getPort(), "us-east-1"))
        .enablePathStyleAccess()
        .build();
```
Fixes the issue. The only change I made from the `S3MockRule.createS3Client()` implementation is using `http` instead of `https`. It doesn't matter if i use `S3MockRule.getPort()` or `S3MockRule.getHttpPort()`, both will work as long as the scheme is `http`.

--- a/server/src/main/java/com/adobe/testing/s3mock/S3MockApplication.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/S3MockApplication.java
@@ -111,6 +111,11 @@ public class S3MockApplication {
   public static final String PROP_HTTP_PORT = "http.port";
 
   /**
+   * Property name for using either HTTPS or HTTP connections
+   */
+  public static final String PROP_SECURE_CONNECTION = "secureConnection";
+
+  /**
    * Property name for enabling the silent mode with logging set at WARN and without banner.
    */
   public static final String PROP_SILENT = "silent";

--- a/server/src/main/java/com/adobe/testing/s3mock/S3MockApplication.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/S3MockApplication.java
@@ -111,7 +111,7 @@ public class S3MockApplication {
   public static final String PROP_HTTP_PORT = "http.port";
 
   /**
-   * Property name for using either HTTPS or HTTP connections
+   * Property name for using either HTTPS or HTTP connections.
    */
   public static final String PROP_SECURE_CONNECTION = "secureConnection";
 

--- a/testsupport/common/src/main/java/com/adobe/testing/s3mock/testsupport/common/S3MockStarter.java
+++ b/testsupport/common/src/main/java/com/adobe/testing/s3mock/testsupport/common/S3MockStarter.java
@@ -85,8 +85,7 @@ public abstract class S3MockStarter {
         .withCredentials(new AWSStaticCredentialsProvider(credentials))
         .withClientConfiguration(
             configureClientToIgnoreInvalidSslCertificates(new ClientConfiguration()))
-        .withEndpointConfiguration(
-            new EndpointConfiguration("https://localhost:" + getPort(), region))
+        .withEndpointConfiguration(getEndpointCongiguration(region))
         .enablePathStyleAccess()
         .build();
   }
@@ -125,6 +124,13 @@ public abstract class S3MockStarter {
             NoopHostnameVerifier.INSTANCE));
 
     return clientConfiguration;
+  }
+
+  protected EndpointConfiguration getEndpointCongiguration(final String region) {
+    boolean isSecureConnection = (boolean) properties.getOrDefault(S3MockApplication.PROP_SECURE_CONNECTION, true);
+    String serviceEndpoint = isSecureConnection ? "https://localhost:" + getPort() : "http://localhost:" + getHttpPort();
+
+    return new EndpointConfiguration(serviceEndpoint, region);
   }
 
   protected void start() {
@@ -208,6 +214,11 @@ public abstract class S3MockStarter {
 
     public BaseBuilder<T> withRootFolder(final String rootFolder) {
       arguments.put(S3MockApplication.PROP_ROOT_DIRECTORY, rootFolder);
+      return this;
+    }
+
+    public BaseBuilder<T> withSecureConnection(final boolean secureConnection) {
+      arguments.put(S3MockApplication.PROP_SECURE_CONNECTION, secureConnection);
       return this;
     }
 

--- a/testsupport/common/src/main/java/com/adobe/testing/s3mock/testsupport/common/S3MockStarter.java
+++ b/testsupport/common/src/main/java/com/adobe/testing/s3mock/testsupport/common/S3MockStarter.java
@@ -127,8 +127,10 @@ public abstract class S3MockStarter {
   }
 
   protected EndpointConfiguration getEndpointCongiguration(final String region) {
-    boolean isSecureConnection = (boolean) properties.getOrDefault(S3MockApplication.PROP_SECURE_CONNECTION, true);
-    String serviceEndpoint = isSecureConnection ? "https://localhost:" + getPort() : "http://localhost:" + getHttpPort();
+    boolean isSecureConnection = (boolean) properties.getOrDefault(
+            S3MockApplication.PROP_SECURE_CONNECTION, true);
+    String serviceEndpoint = isSecureConnection ? "https://localhost:" + getPort()
+            : "http://localhost:" + getHttpPort();
 
     return new EndpointConfiguration(serviceEndpoint, region);
   }

--- a/testsupport/junit5/src/test/java/com/adobe/testing/s3mock/junit5/S3MockExtensionProgrammaticTest.java
+++ b/testsupport/junit5/src/test/java/com/adobe/testing/s3mock/junit5/S3MockExtensionProgrammaticTest.java
@@ -33,7 +33,8 @@ import org.junit.jupiter.api.extension.RegisterExtension;
  */
 public class S3MockExtensionProgrammaticTest {
   @RegisterExtension
-  static final S3MockExtension S3_MOCK = S3MockExtension.builder().silent().withSecureConnection(false).build();
+  static final S3MockExtension S3_MOCK = S3MockExtension.builder().silent()
+          .withSecureConnection(false).build();
 
   private static final String BUCKET_NAME = "mydemotestbucket";
   private static final String UPLOAD_FILE_NAME = "src/test/resources/sampleFile.txt";

--- a/testsupport/junit5/src/test/java/com/adobe/testing/s3mock/junit5/S3MockExtensionProgrammaticTest.java
+++ b/testsupport/junit5/src/test/java/com/adobe/testing/s3mock/junit5/S3MockExtensionProgrammaticTest.java
@@ -33,7 +33,7 @@ import org.junit.jupiter.api.extension.RegisterExtension;
  */
 public class S3MockExtensionProgrammaticTest {
   @RegisterExtension
-  static final S3MockExtension S3_MOCK = S3MockExtension.builder().silent().build();
+  static final S3MockExtension S3_MOCK = S3MockExtension.builder().silent().withSecureConnection(false).build();
 
   private static final String BUCKET_NAME = "mydemotestbucket";
   private static final String UPLOAD_FILE_NAME = "src/test/resources/sampleFile.txt";

